### DISCUSSION TITLE
SPI Support for ssd20xd (Infinity2m)

### DISCUSSION THREAD

**Discussion by AhmadPanogari** [linux-chenxing/linux-chenxing.org#61] [General]
Hello, I'm currently working on the development of a board with a **SSD202D** chip

but right now I need an **SPI Interface** to interconnect with other devices.

can you help me to present the **SPI Interface** so that I can use it?

I'm currently using the **DongshanPi One** development board

maybe you can also help to complete the interface available in the development board.

Thanks very much.

**Comment by fifteenhex** [comment]
SPI is already supported. What you need to do is enable the correct SPI controller and the pinctrl settings to have it present on the right pins. Can you tell me which pins you need it on? I guess you are using the baseboard?

**Reply by AhmadPanogari** [reply]
Yes, i use baseboard of DongshanPi One

which is circled the SPI Pin that I will use

![InkedDongshanPI-ONE-Sigmstar-SSD202D_LI](https://user-images.githubusercontent.com/17040103/168266273-a37ad036-c6a1-48f2-9a53-9de372f1ab6c.jpg)

From the shematic

![image](https://user-images.githubusercontent.com/17040103/168266949-04467886-e086-4b2e-9b26-fbcf67003ae6.png)

**Reply by fifteenhex** [reply]
It looks like we need a little fix to the kernel to make it work.. testing it now. :)

**Reply by AhmadPanogari** [reply]
can you also help to get GPIO, SPI_A, SPI_B, I2C_A, I2C_B and UART_A working properly.

and at least it can be seen in the /dev/ directory

**Reply by fifteenhex** [reply]
Can you try mstar_v5_18_rebase(https://github.com/linux-chenxing/linux/commit/a2a8b9241e004a041320c6ada36dd2290c0a7175)?
You'll need to add your display node here: https://github.com/linux-chenxing/linux/blob/27775873cbf64feaa152cf5fcf1e5bf37e8ce38d/arch/arm/boot/dts/mstar-infinity2m-ssd202d-100ask-dongshanpione-carrier.dts#L26

As long as you don't mind wiring stuff up and testing it I will get those other connections working. I have checked the SPI with a scope but not wired anything up.

**Reply by AhmadPanogari** [reply]
Ok I will connect the display to SPI Interface

**Reply by AhmadPanogari** [reply]
This my display node

![image](https://user-images.githubusercontent.com/17040103/168278511-c8c03e22-4e81-4861-85cf-d66c3fa67e26.png)

I want to use Display ILI9341 240x320 3.2"

is this configuration correct?

**Reply by fifteenhex** [reply]
It looks ok to me. I think the only way to know for sure is to try it. I don't think you'll get 33fps out of it right now as there is no DMA for the SPI.

**Reply by AhmadPanogari** [reply]
![image](https://user-images.githubusercontent.com/17040103/168309245-6a639f2f-092d-442e-9b08-a40a12a4e134.png)

This is log of boot process.

but i cannot see fb0 in /dev/ dictionary

any have missing or error about this?

**Reply by fifteenhex** [reply]
It looks like it's complaining the bus width is not set. Can you find where the message comes from (which line) and see if that is a fatal error?

**Reply by AhmadPanogari** [reply]
[    1.658335] SPI driver fb_ili9340 has no spi_device_id for ilitek,ili9340
[    1.665421] SPI driver fb_ili9341 has no spi_device_id for ilitek,ili9341
[    1.672353] fb_ili9341 spi0.0: fbtft_property_value: buswidth = 8
[    1.678480] fb_ili9341 spi0.0: fbtft_property_value: debug = 4
[    1.684374] fb_ili9341 spi0.0: fbtft_property_value: rotate = 90
[    1.690419] fb_ili9341 spi0.0: fbtft_property_value: fps = 33
[    1.696589] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'reset' GPIO
[    1.702968] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'dc' GPIO
[    1.709029] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'rd' GPIO
[    1.715105] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'wr' GPIO
[    1.721159] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'cs' GPIO
[    1.727234] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'latch' GPIO
[    1.733574] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.739629] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.745783] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.751923] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.758001] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.764163] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.770304] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.776381] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.782533] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.795452] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.808309] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.814471] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.820614] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.826689] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.832845] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.838985] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.845050] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.851192] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.857345] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.863411] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.869552] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.875714] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.881770] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.887930] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.894086] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.900141] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.906291] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.912452] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.918508] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.924659] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.930799] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.936867] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.943027] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.949167] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.955237] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.961379] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.967534] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.973604] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.979744] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    1.985904] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    1.991960] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    1.998114] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    2.004290] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    2.010349] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    2.016512] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    2.022675] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'db' GPIO
[    2.032503] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'led' GPIO
[    2.042636] fb_ili9341 spi0.0: fbtft_request_one_gpio: 'aux' GPIO
[    2.789738] fb_ili9341 spi0.0: Display update: 311 kB/s, fps=0
[    2.804482] graphics fb0: fb_ili9341 frame buffer, 320x240, 150 KiB video memory, 16 KiB buffer memory, 
[    2.816487] SPI driver fb_ssd1306 has no spi_device_id for solomon,ssd1306
[    2.823653] SPI driver fb_st7789v has no spi_device_id for sitronix,st7789v
[    2.849126] SPI driver st-accel-spi has no spi_device_id for st,lis302dl-spi
[    2.856255] SPI driver st-accel-spi has no spi_device_id for st,lis3lv02dl-accel
[    2.863699] SPI driver st-accel-spi has no spi_device_id for st,lis3dh-accel
[    2.870775] SPI driver st-accel-spi has no spi_device_id for st,lsm330d-accel
[    2.877949] SPI driver st-accel-spi has no spi_device_id for st,lsm330dl-accel
[    2.885214] SPI driver st-accel-spi has no spi_device_id for st,lsm330dlc-accel
[    2.899518] SPI driver st-accel-spi has no spi_device_id for st,lis331dlh-accel
[    2.906865] SPI driver st-accel-spi has no spi_device_id for st,lsm330-accel
[    2.913960] SPI driver st-accel-spi has no spi_device_id for st,lsm303agr-accel
[    2.921296] SPI driver st-accel-spi has no spi_device_id for st,lis2dh12-accel
[    2.928555] SPI driver st-accel-spi has no spi_device_id for st,lng2dm-accel
[    2.935643] SPI driver st-accel-spi has no spi_device_id for st,h3lis331dl-accel
[    2.943079] SPI driver st-accel-spi has no spi_device_id for st,lis331dl-accel
[    3.466701] spi-nand spi1.0: XINCUN SPI NAND was found.
[    3.471978] spi-nand spi1.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 128
[    3.483805] mtdblock: MTD device 'spi1.0' is NAND, please consider using UBI block devices instead.
[    3.506736] 1 fixed-partitions partitions found on MTD device spi1.0
[    3.513214] Creating 1 MTD partitions on "spi1.0":

this log after i add a **buswidth = <8>** in .dtb

but the display just blink and display white screen.

**Reply by fifteenhex** [reply]
I'm not sure what the buswidth parameter should be.. does `cat /dev/urandom > /dev/fb0` do anything?

**Reply by AhmadPanogari** [reply]
![image](https://user-images.githubusercontent.com/17040103/168468739-b498e840-9b23-4fc9-823a-a5dcc1f235e9.png)

This is the response
